### PR TITLE
Speedup by calling Buffer.concat less often.

### PR DESCRIPTION
**Pull request by tylerneylon**
Hello @evansolomon, 

Please review the following commits I made in branch 'tyler-less-concat-calls'.

23afd5aacac4431e9f43a68ae9187914799f4e9b (2013-09-06 04:36:59 +0000)
Speedup by calling Buffer.concat less often.

R=@evansolomon

### CODE REVIEW

File: lib/Queue.js [hunk 1]
```diff
@@ -24,7 +24,8 @@ module.exports = S3Queue;
  */
 S3Queue.prototype.reset = function () {
   this.emit('reset');
-  this.buffer = new Buffer(0);
+  this.buffer = [];
```
Comment (evansolomon) [inline lib/Queue.js]: I don't think this should be called `buffer` (little "b") if it's not an instance of `Buffer` (big "B").  How about `chunks`?

File: lib/Queue.js [hunk 2]
```diff
@@ -24,7 +24,8 @@ module.exports = S3Queue;
...
+  this.bufferLen = 0;
```
Comment (evansolomon) [inline lib/Queue.js]: "Length" is a little confusing here since the underlying data is an array, but this _isn't_ the array's length.  How about `size`?

File: lib/Queue.js [hunk 3]
```diff
@@ -55,7 +56,9 @@ S3Queue.prototype.full = function () {
  * @param {Buffer} chunk Data to add to the queue.
  */
 S3Queue.prototype.push = function (chunk) {
-  this.buffer = Buffer.concat([this.buffer, chunk]);
+  if (typeof chunk === 'string') chunk = Buffer(chunk);
```
Comment (evansolomon) [inline lib/Queue.js]: The chunks provided by `S3Stream` will always be instances of `Buffer`. Also the `chunk` param is specified as a `Buffer` in the doc block.

**Comment by evansolomon** [comment]
This does seem to be faster.  I did a simple benchmark to see how this affects performance. I picked three files, ~1 mb, ~27 mb, and ~290 mb, and piped them to s3.

tl;dr -- As the data gets bigger, the resources saved increases significantly. The actual time to upload the object doesn't necessarily get any faster, but that is mostly driven by time spent in network requests.

Here's the benchmark script:

``` javascript
var fs = require('fs');
var Canoe = require('./index');
var AWS = require('aws-sdk');

var s3 = new AWS.S3();
var canoe = new Canoe(s3);

fs.createReadStream('/path/to/some-file')
.pipe(canoe.createWriteStream({
  Bucket: 's3-bucket',
  Key: 'some-file'
}));
```

I used `time` and ran a few uploads.  My assumption is that we should ignore the total time (since it's influenced largely by network requests), and just pay attention to the user and system time since that's where our actual resources are being used.

### 1 mb file

#### Current master

```
$ time node speedtest.js
node speedtest.js  0.33s user 0.07s system 9% cpu 4.228 total
```

#### Less concats

```
$ time node speedtest.js
node speedtest.js  0.34s user 0.06s system 10% cpu 3.862 total
```

### 27 mb file

#### Current master

```
$ time node speedtest.js
node speedtest.js  3.47s user 1.34s system 9% cpu 50.568 total
```

#### Less concats

```
$ time node speedtest.js
node speedtest.js  2.57s user 0.65s system 6% cpu 48.282 total
```

### 290 mb file

#### Current master

```
$ time node speedtest.js
node speedtest.js  30.18s user 13.22s system 8% cpu 8:35.32 total
```

#### Less concats

```
$ time node speedtest.js
node speedtest.js  25.31s user 6.22s system 6% cpu 8:38.18 total
```

**Comment by dpup** [comment]
I believe concat allocates a new object. So this makes sense right?

-- Dan

**Comment by evansolomon** [comment]
> I believe concat allocates a new object. So this makes sense right?

Yes, it creates a new 0 length Buffer than uses `buffer.copy()` for each item in the list.

**Comment by evansolomon** [comment]
I started working on some more detailed benchmarks, this is definitely a nice improvement.

#### Master

```
Running benchmark S3Queue: Create instance [benchmark/queue/create.js]...
>> S3Queue: Create instance x 1,058,932 ops/sec ±5.39% (85 runs sampled)

Running suite S3Queue: Push many items (~10 MB) [benchmark/queue/push-many-items.js]...
>> Big items x 4.33 ops/sec ±6.88% (15 runs sampled)
>> Medium items x 0.38 ops/sec ±11.00% (5 runs sampled)
>> Small items x 0.04 ops/sec ±10.32% (5 runs sampled)
```

#### With this PR

```
Running benchmark S3Queue: Create instance [benchmark/queue/create.js]...
>> S3Queue: Create instance x 3,797,069 ops/sec ±2.83% (94 runs sampled)

Running suite S3Queue: Push many items (~10 MB) [benchmark/queue/push-many-items.js]...
>> Big items x 243 ops/sec ±7.98% (37 runs sampled)
>> Medium items x 204 ops/sec ±8.95% (46 runs sampled)
>> Small items x 92.27 ops/sec ±5.45% (63 runs sampled)
```

**Comment by tylerneylon** [comment]
Cool. I also ran some other speed tests with good results:

```
var fs = require('fs');
var Canoe = require('./index');
var AWS = require('aws-sdk');

var s3 = new AWS.S3();
var canoe = new Canoe(s3);

var writer = canoe.createWriteStream({
  Bucket: 'product-science',
  Key: 'canoe-test'
});

for (var i = 0; i < 512 * 1024; i++) {
  writer.write('a');
}

writer.end();
```

This is writing 512K of a's (they're from Oakland).

speedup-branch: 

```
real    0m5.967s
user    0m4.696s
sys     0m0.552s
```

master:

```
real    3m21.779s
user    3m15.960s
sys     0m5.288s
```

I originally got lackluster results until I realized I was setting my AWS keys as environment variables (feels like it's easy to not notice that mistake - failed silently).

**Comment by tylerneylon** [comment]
@evansolomon, take one more look? I've addressed the previous comments.

Thanks for taking some time to generate some useful benchmarks on this, too.
